libobs/obs-source.c

```diff
@@ -3579,11 +3586,17 @@ cache_video(struct obs_source *source, const struct obs_source_frame *frame)
 
 	pthread_mutex_lock(&source->async_mutex);
 
-	if (source->async_frames.num >= MAX_ASYNC_FRAMES) {
-		free_async_cache(source);
+	if ((source->async_frames.num + 1) >= MAX_ASYNC_FRAMES) {
+		/* Reset timestamp and throw away all old frames without
+		 * freeing the cache's memory. */
 		source->last_frame_ts = 0;
-		pthread_mutex_unlock(&source->async_mutex);
-		return NULL;
+		source->async_last_clean = obs_get_video_frame_time();
+
+		da_clear(source->async_frames);
+		for (size_t i = 0; i < source->async_cache.num; i++) {
+			source->async_cache.array[i].used = false;
+			source->async_cache.array[i].unused_count = 0;
+		}
```

Comment:
The problem with doing this is that the point of the `used` member is that the frame is in use somewhere. For example, an async delay filter will use frames while they're in the cache.

How can one reproduce the original performance issue where it releases all the frames?

Reply:
Pause OBS in a debugger for a bit while as async source is feeding it frames, then continue.

Reply:
Well I meant like a real-world example of how this ends up happening. Because the max queued async frame count should be fairly big.

Also the async delay filter case is still a situation we probably have to think about. You could end up overwriting frames that are still technically referenced inside of an async delay filter. That's technically why the old frames are just released. The idea is "this situation *shouldn't* happen very often, but if it does, then just release the frames because they're probably in use somewhere." So that situation might need to be taken into account. I'm not ruling out this code as an optimization, I'm mostly just noting that's the situation that you may end up in if you force frames to an unused state.

Reply:
As far as real-world examples go, network streams are on thing. e.g. if you run into buffering or stalls and then a ton of frames come in at once. I'll take a look at the async delay filter case, but I fear that taking that into account may complicate things to the point where it's not really worth it to work around that in this PR.

Reply:
Yea, I know. I'm not sure what to do either.

Reply:
This was mostly spawned out of a discussion with R1CH and his complaints about memory churn when we are in a catch-up scenario, so it's not super important. I'm sure we can improve upon the current situation, but I don't think it's worth keeping this PR around for now.